File: Data Structures & Algorithms/binary-tree-right-side-view/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    List<Integer> result = new ArrayList<>();
    public List<Integer> rightSideView(TreeNode root) {
        visit(root,0);
        return result;
    }

    public void visit (TreeNode node, int height){
        if(node == null){
            return;
        }

        if(result.size() == height){
            result.add(node.val);
        }
        
        height++;
        visit(node.right, height);
        visit(node.left, height);
    }
}

File: Data Structures & Algorithms/serialize-and-deserialize-binary-tree/submission-1.cpp
class Codec {
    struct BitWriter {
        string s;
        int bit = 0;

        void writeBit(int x) {
            if (bit == 0) s.push_back(0);
            if (x) s.back() |= char(1 << bit);
            bit = (bit + 1) & 7;
        }

        void writeInt(int x) {
            unsigned int v = (unsigned int)x;
            for (int i = 0; i < 32; i++) {
                writeBit((v >> i) & 1);
            }
        }
    };

    struct BitReader {
        const string& s;
        int idx = 0;
        int bit = 0;

        BitReader(const string& data) : s(data) {}

        int readBit() {
            int x = ((unsigned char)s[idx] >> bit) & 1;
            bit++;
            if (bit == 8) {
                bit = 0;
                idx++;
            }
            return x;
        }

        int readInt() {
            unsigned int v = 0;
            for (int i = 0; i < 32; i++) {
                if (readBit()) v |= (1u << i);
            }
            return (int)v;
        }
    };

    void writeTree(TreeNode* node, BitWriter& bw) {
        if (!node) {
            bw.writeBit(0);
            return;
        }

        bw.writeBit(1);
        bw.writeInt(node->val);
        writeTree(node->left, bw);
        writeTree(node->right, bw);
    }

    TreeNode* readTree(BitReader& br) {
        if (!br.readBit()) return nullptr;

        TreeNode* node = new TreeNode(br.readInt());
        node->left = readTree(br);
        node->right = readTree(br);

        return node;
    }

public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        BitWriter bw;
        writeTree(root, bw);
        return bw.s;
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        BitReader br(data);
        return readTree(br);
    }
};